### PR TITLE
Added image synthesizing using poisson blending

### PR DESCRIPTION
**Pull request by augustak**
- Initial image synth, additional error handling in PDF synth
- Updated image synth
- Refactored code, added jitter

### CODE REVIEW

File: synthetic/image/parser.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    for i in range(num_outputs_per_document):
+        if not (_out_path(i, '.jpeg').exists() and _out_path(i, '.json').exists()):
+            k_to_process.append(i)
+
+    if not k_to_process:
+        raise AlreadyProcessed(f'Already processed {image_file} {json_file}')
+
+    ground_truth = json.loads(json_file.read_text())
+    from_size, to_size = font_size_range
+    fonts = {i: ImageFont.truetype(str(font_path), i) for i in range(to_size, from_size, -1)}
```
Comment (magnurud) [inline synthetic/image/parser.py:49]: why is this done in reversed order?

**Review by magnurud — COMMENTED** [review]
Should any of the functions in poisson_blending.py be reviewed as new code? Or is the intention to keep it as similar to the original code as possible?

File: synthetic/image/parser.py (hunk 1 above)
Comment (augustak) [reply synthetic/image/parser.py:49]: We wish to try fit the biggest font size first

File: synthetic/image/poisson_blending.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    image_array[ymb:yMb, xmb:xMb, :] = image_ret
+    return image_array
+
+
+def adjust_box(edginess, bounding_box):
+    r = 0.1
+    xmin, ymin, xmax, ymax = bounding_box
+    dx, dy = int(xmax - xmin), int(ymax - ymin)
+    dx, dy = int(r * sqrt(dy * dx)), int(r * sqrt(dy * dx))
+    dx, dy = max(dx, 5), max(dy, 5)
```
Comment (magnurud) [inline synthetic/image/poisson_blending.py:174]: isn't dx and dy the same number here?

File: synthetic/image/poisson_blending.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    D = max(2, int(np.sqrt(dx * dy) / 10))
+    for possible_box in itertools.product(
+        range(xmin - dx, xmin + dxh + 1),
+        range(ymin - dy, ymin + dyh + 1),
+        range(xmax - dxh, xmax + dx + 1),
+        range(ymax - dxh, ymax + dy + 1),
+    ):
+        xm, ym, xM, yM = possible_box
+        e1 = edginess[ym - D:yM + D + 1, xm - D:xM + D + 1]
+        e2 = edginess[ym + D:yM - D + 1, xm + D:xM - D + 1]
```
Comment (magnurud) [inline synthetic/image/poisson_blending.py:188]: Do we suddenly have y-index as the first index here?

File: synthetic/image/poisson_blending.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        E = np.sum(e1) - np.sum(e2)
+        if E < min_val:
```
Comment (magnurud) [inline synthetic/image/poisson_blending.py:190]: This is kind of like checking the gradient of the box, the lesser the gradient the better?

File: synthetic/image/poisson_blending.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+# TODO: consider doing a check here to ensure boxes do not overlap
+def blend_image(image: Image, bounding_boxes: list, fonts: dict, max_size: int):
+    # Resize if image is too large. Large images will have a performance impact
+    if image.size[0] * image.size[1] > max_size:
+        s = math.sqrt(image.size[0] * image.size[1] / max_size)
+        resized_image = image.resize(tuple(map(int, (image.size[0] / s, image.size[1] / s))))
+        return blend_image(resized_image, bounding_boxes, fonts, max_size).resize(image.size)
+
+    bounding_boxes = [(rel_to_abs(bounding_box, image.size), text) for bounding_box, text in bounding_boxes]
+    image_bw_arr = np.asarray(image.convert('L'))
```
Comment (magnurud) [inline synthetic/image/poisson_blending.py:215]: bw?

**Review by magnurud — CHANGES_REQUESTED** [review]

File: synthetic/image/poisson_blending.py (hunk 5 above)
Comment (augustak) [reply synthetic/image/poisson_blending.py:215]: black-white, renamed to binary

File: synthetic/image/poisson_blending.py [hunk 6]
```diff
@@ -15,77 +15,71 @@
...
+        if image_mask[y, x]:
+            index = x + y * region_size[1]
+            A[index, index] = 4
+            if index + 1 < np.prod(region_size):
+                A[index, index + 1] = -1
+            if index - 1 >= 0:
+                A[index, index - 1] = -1
+            if index+region_size[1] < np.prod(region_size):
+                A[index, index + region_size[1]] = -1
+            if index-region_size[1] >= 0:
```
Comment (magnurud) [inline synthetic/image/poisson_blending.py:60]: Give the operators some space for breathing
```suggestion
            if index + region_size[1] < np.prod(region_size):
                A[index, index + region_size[1]] = -1
            if index - region_size[1] >= 0:
```

File: synthetic/image/poisson_blending.py [hunk 7]
```diff
@@ -135,44 +133,51 @@ def text_to_image(text, image_size, fonts_by_size, bg_color, fg_color, angle):
...
+        angle=angle,
+    )
     image_ret = blend(image_target, np.asarray(image_source), image_mask)
 
-    image_array[ymb:yMb, xmb:xMb, :] = image_ret
+    image_array[ymin_b:ymax_b, xmin_b:xmax_b, :] = image_ret
     return image_array
 
 
 def adjust_box(edginess, bounding_box):
```
Comment (magnurud) [inline synthetic/image/poisson_blending.py:174]:
```suggestion
def adjust_box(edginess, bounding_box):
"""
Find the best bounding box by minimizing the difference in edges on the boundary
"""
```

**Review by magnurud — CHANGES_REQUESTED** [review]

**Review by magnurud — APPROVED** [review]

**Pull request merged**